### DISCUSSION TITLE
How to locate elements built with flutter and canvas

### DISCUSSION THREAD

**Discussion by dhamo-pk** [cypress-io/cypress#26177] [Questions and Help] [answered]
I am trying to click on the Privacy policy on this web site -  https://m.exeter.ac.uk/acceptCookies

with this code:
 ```
cy.visit('https://m.exeter.ac.uk/');
    cy.get('flt-glass-pane')
      .should('be.visible')
      .shadow()
      .find('canvas')
      .scrollIntoView()
      .contains('privacy')
      .should('be.visible');
```
But I was not able to succeed in my attempt. Please give some hints how I should go about it? 
Error:
`Timed out retrying after 4000ms: Expected to find content: 'privacy' within the element: <canvas> but never did.`

**Comment by portyu9** [comment, marked as answer]
Hi @dhamo-pk 

`cy.contains()` cannot find text that is **drawn into a `<canvas>`**.

This:

```js
.find('canvas')
.contains('privacy')
```

asks Cypress to find a DOM descendant containing the text `privacy`. But text painted into a canvas is pixels, not DOM text, so there is no descendant element for Cypress to locate.

For a Flutter web application you control, the preferred solution is to expose proper **Flutter semantics/accessibility information**. Flutter can translate its semantics tree into an accessible HTML DOM.

For example, current Flutter supports enabling semantics programmatically:

```dart
import 'package:flutter/semantics.dart';

void main() {
  runApp(const MyApp());

  if (kIsWeb) {
    SemanticsBinding.instance.ensureSemantics();
  }
}
```

Also make sure interactive widgets have meaningful semantics/roles/labels. Cypress can then interact with the semantic DOM rather than trying to inspect canvas pixels.

If you **do not control the Flutter application**, Cypress cannot turn arbitrary canvas text into DOM elements. You could technically click coordinates on the canvas, but that is very brittle because it depends on screen size, layout and rendering.

So the solution is not another Cypress selector: either expose an accessible semantic layer from the Flutter application, or use a testing approach that can interact with Flutter at the widget/semantics level.

Please mark this answer as accepted if it helped, thank you.
